### ISSUE TITLE
Action profile installation fails with "serious errors" due to undefined behavior in Judy library

### ISSUE THREAD

**Issue by smolkaj** [p4lang/behavioral-model#1001]
I noticed that BMv2 seems to occasionally replicate group members of  `action_profile_action_set`s. Concretely, when installing the following entry (along with several other entries),
```
table_id: 33554499
match {
  field_id: 1
  exact {
    value: "group-4294934549"
  }
}
action {
  action_profile_action_set {
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-8"
        }
      }
      weight: 15
      watch_port: "\r"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-9"
        }
      }
      weight: 14
      watch_port: "\023"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-11"
        }
      }
      weight: 14
      watch_port: "\027"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-12"
        }
      }
      weight: 14
      watch_port: "\033"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-21"
        }
      }
      weight: 14
      watch_port: "-"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-23"
        }
      }
      weight: 14
      watch_port: "1"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-24"
        }
      }
      weight: 14
      watch_port: "5"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-25"
        }
      }
      weight: 14
      watch_port: "7"
    }
    action_profile_actions {
      action {
        action_id: 16777221
        params {
          param_id: 1
          value: "nexthop-27"
        }
      }
      weight: 14
      watch_port: "="
    }
  }
}
```
and then injecting a packet that hits this table entry, I see the following in the BMv2 log:
```
[12:14:40.301] [bmv2] [T] [thread 1142658] [1.0] [cxt 0] Applying table 'ingress.routing.wcmp_group_table'
[12:14:40.301] [bmv2] [D] [thread 1142658] [1.0] [cxt 0] Looking up key:
* wcmp_group_id       : 000000000000000067726f75702d34323934393334353439

[12:14:40.301] [bmv2] [D] [thread 1142658] [1.0] [cxt 0] Choosing member 1046 from group 20
[12:14:40.301] [bmv2] [D] [thread 1142658] [1.0] [cxt 0] Table 'ingress.routing.wcmp_group_table': hit with handle 20
[12:14:40.302] [bmv2] [D] [thread 1142658] [1.0] [cxt 0] Dumping entry 20
Match key:
* wcmp_group_id       : EXACT     000000000000000067726f75702d34323934393334353439
Index: group(20)
Group members:
  mbr 1045: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1046: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1047: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1048: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1049: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1050: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1051: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1052: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1053: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1054: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1055: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1056: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1057: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1058: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1059: ingress.routing.set_nexthop_id - 6e657874686f702d38,
  mbr 1060: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1061: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1062: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1063: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1064: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1065: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1066: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1067: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1068: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1069: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1070: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1071: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1072: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1073: ingress.routing.set_nexthop_id - 6e657874686f702d39,
  mbr 1074: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1075: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1076: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1077: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1078: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1079: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1080: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1081: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1082: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1083: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1084: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1085: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1086: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1087: ingress.routing.set_nexthop_id - 6e657874686f702d3131,
  mbr 1088: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1089: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1090: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1091: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1092: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1093: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1094: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1095: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1096: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1097: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1098: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1099: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1100: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1101: ingress.routing.set_nexthop_id - 6e657874686f702d3132,
  mbr 1102: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1103: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1104: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1105: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1106: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1107: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1108: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1109: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1110: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1111: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1112: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1113: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1114: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1115: ingress.routing.set_nexthop_id - 6e657874686f702d3231,
  mbr 1116: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1117: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1118: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1119: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1120: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1121: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1122: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1123: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1124: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1125: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1126: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1127: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1128: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1129: ingress.routing.set_nexthop_id - 6e657874686f702d3233,
  mbr 1130: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1131: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1132: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1133: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1134: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1135: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1136: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1137: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1138: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1139: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1140: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1141: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1142: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1143: ingress.routing.set_nexthop_id - 6e657874686f702d3234,
  mbr 1144: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1145: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1146: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1147: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1148: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1149: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1150: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1151: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1152: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1153: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1154: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1155: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1156: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1157: ingress.routing.set_nexthop_id - 6e657874686f702d3235,
  mbr 1158: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1159: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1160: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1161: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1162: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1163: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1164: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1165: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1166: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1167: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1168: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1169: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1170: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
  mbr 1171: ingress.routing.set_nexthop_id - 6e657874686f702d3237,
```
Is this expected behavior or a bug? It seems that the number of duplicates is equal to the weight, which indicates this may be intentional.

However, when I set all weights to 1, I get the a programming error such as the following: `INTERNAL: 9 serious errors when encountered during cleanup; you may need to reboot the device`.

**Comment by antoninbas** [comment]
Repeated members are expected: that's how bmv2 implements weights. I believe this is pretty common, which is why the definition of `size` for an action selector in the P4Runtime spec is:

> size, an int64 representing the maximum number of member entries that the Action Profile can hold, or, in the case of an Action Selector, the maximum sum of all member weights across all selector groups.

Regarding the error when you set the weights to 1. I'm not sure why it would fail with weights of 1 but not weights greater than 1. Do you have a sequence of P4Runtime messages to reproduce the issue?
BTW, last time you ran into a similar issue, I updated the bmv2 log messages to include more information (the exact error code): https://github.com/p4lang/behavioral-model/pull/979. Is there anything useful in the logs?

**Comment by smolkaj** [comment]
Thanks Antonin for the platinum support :) 

>  Is there anything useful in the logs?
```
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 258 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 259 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 260 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 261 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 262 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 263 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 264 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 265 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 266 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 267 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 268 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Added member 269 to action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.846] [bmv2] [D] [thread 1169440] Created group 6 in action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to add member 258 to group 6 in action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error when adding member to group on target
[16:08:20.847] [bmv2] [D] [thread 1169440] Removed group 6 from action profile 'ingress.routing.wcmp_group_selector'
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 269 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 268 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 267 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 266 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 265 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 264 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 263 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.847] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.847] [bmv2] [E] [thread 1169440] Error when trying to remove member 262 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.848] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.848] [bmv2] [E] [thread 1169440] Error when trying to remove member 261 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.848] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.848] [bmv2] [E] [thread 1169440] Error when trying to remove member 260 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.848] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.848] [bmv2] [E] [thread 1169440] Error when trying to remove member 259 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.848] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.848] [bmv2] [E] [thread 1169440] Error when trying to remove member 258 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:08:20.848] [bmv2] [E] [thread 1169440] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:08:20.848] [bmv2] [E] [thread 1169440] [P4Runtime] 12 serious errors when encountered during cleanup; you may need to reboot the device
```

> Do you have a sequence of P4Runtime messages to reproduce the issue?

It doesn't happen if I only install the entry from my first comment. I have a huge sequence that reproduces the issue deterministically, but let me try to shrink it to something workable.

**Comment by smolkaj** [comment]
The following 6 table entries installed in sequence reproduce the issue reliably:
[table_entries.txt](https://github.com/p4lang/behavioral-model/files/6444730/table_entries.txt)

I will also try to get you the sequence of full messages, but wanted to get the entries to you first in case that is already helpful.

**Comment by antoninbas** [comment]
Sorry, do I need the full sequence or just these 6 table entries? Could I have the P4 table definition with the action selector too?

**Comment by smolkaj** [comment]
I'm reproducing the problem by installing only those 6 table entries (and maybe some config push before that).
Here is the table:
```
control routing() {
  bit<192> wcmp_group_id_value;
  bit<16> wcmp_selector_input;

  @id(0x01000005)
  action set_nexthop_id(@id(1) bit<192> nexthop_id) { }

  @max_group_size(1024)
  action_selector(HashAlgorithm.identity, 65536, 16) wcmp_group_selector;

  @id(0x02000043)
  @oneshot()
  table wcmp_group_table {
    key = {
      wcmp_group_id_value : exact @id(1) @name("wcmp_group_id");
      wcmp_selector_input : selector;
    }
    actions = {
      @proto_id(1) set_nexthop_id;
      @defaultonly NoAction;
    }
    const default_action = NoAction;
    implementation = wcmp_group_selector;
    size = WCMP_GROUP_TABLE_MINIMUM_GUARANTEED_SIZE;
  }

  apply { wcmp_group_table.apply(); }
}
```

**Comment by smolkaj** [comment]
Here is a complete, minimal program that reproduces the issue together with the 6 table entries from my earlier comment:
[program.p4.txt](https://github.com/p4lang/behavioral-model/files/6444782/program.p4.txt)

**Comment by smolkaj** [comment]
Let me know if you still want a complete list of all messages.
They should be:
- arbitration request
- SetBmv2ForwardingPipelineConfig
- INSERT the 6 table entries one at a time (6 batches).

**Comment by antoninbas** [comment]
I tried to play the table entries you provided with p4runtime-shell and couldn't reproduce the failure.
I tried a few combinations: individual batches, one batch with all entries, no watch port, with watch port. Everything worked well.
Note that when using individual batches, all batches / WriteRequests were sent sequentially, with a 1 second pause in between them (I had to type a new command in the p4runtime-shell). I don't know if in your case batches are in parallel?

I was wondering if there was anything specific with your bmv2 build or if it's completely standard? Anything interesting happening with the libJudy dependency (which manages member handles)?

Another interesting thing. In my case the maximum member handle I observe is 257:
```
[19:24:11.707] [bmv2] [D] [thread 1870459] Added member 257 to group 5 in action profile 'IngressI.wcmp_group_selector'
```
With the logs you shared above (https://github.com/p4lang/behavioral-model/issues/1001#issuecomment-834853101), I see that problems start happening with member handle 258? Is that a coincidence. Do the logs above correspond to the table entries you shared with me?

**Comment by antoninbas** [comment]
I see that you shared stdout above.
It's really strange. I have the exact same logs as you, up until this part which is completely missing for me:
```
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 258 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 259 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 260 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 261 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 262 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 263 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 264 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 265 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 266 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.996] [bmv2] [D] [thread 1218263] Added member 267 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.997] [bmv2] [D] [thread 1218263] Added member 268 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.997] [bmv2] [D] [thread 1218263] Added member 269 to action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.997] [bmv2] [D] [thread 1218263] Created group 6 in action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to add member 258 to group 6 in action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.997] [bmv2] [E] [thread 1218263] [P4Runtime] Error when adding member to group on target
[16:26:42.997] [bmv2] [D] [thread 1218263] Removed group 6 from action profile 'ingress.routing.wcmp_group_selector'
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to remove member 269 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.997] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to remove member 268 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.997] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to remove member 267 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.997] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to remove member 266 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.997] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to remove member 265 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.997] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.997] [bmv2] [E] [thread 1218263] Error when trying to remove member 264 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] Error when trying to remove member 263 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] Error when trying to remove member 262 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] Error when trying to remove member 261 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] Error when trying to remove member 260 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] Error when trying to remove member 259 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] Error when trying to remove member 258 from action profile 'ingress.routing.wcmp_group_selector': INVALID_MBR_HANDLE
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] Error encountered when cleaning up action profile member created by one-shot indirect table programming. This is a serious error and you may need to reboot the system
[16:26:42.998] [bmv2] [E] [thread 1218263] [P4Runtime] 12 serious errors when encountered during cleanup; you may need to reboot the device
```

**Comment by antoninbas** [comment]
There are 6 table entries, so 6 groups. But the logs show that simple_switch_grpc is trying to create 7 groups, and fails on the 7th one. So that's probably connected to the issue. Still have no clue why it's logging that the member was created successfully, but then complains that it doesn't exist, with nothing interesting happening in between...

**Comment by antoninbas** [comment]
Your logs show 2 different threads performing updates, while mine only have 1 thread. They seem to alternate. I don't know if that means you are sending batches in parallel. Probably...

**Comment by antoninbas** [comment]
BTW, I noticed this in the log:
```
[16:26:42.880] [bmv2] [W] [thread 1218263] [P4Runtime] p4::tmp::P4DeviceConfig is deprecated
```
I don't see how it can be related to the issue, but it may be worth to avoid using `P4DeviceConfig` altogether since there is a slightly different code path in the P4Runtime server code. It's not needed, the bmv2 JSON file can be passed directly as the `p4_device_config` binary string.

**Comment by smolkaj** [comment]
> There are 6 table entries, so 6 groups. But the logs show that simple_switch_grpc is trying to create 7 groups, and fails on the 7th one. 

Apologies, I indeed didn't send you the crucial 7th entry 🤦 . Here we go.
[table_entries.pb.txt](https://github.com/p4lang/behavioral-model/files/6454389/table_entries.pb.txt)

**Comment by smolkaj** [comment]
Very good suggestion. Jackpot!

```
third_party/judy/src/JudyCommon/JudyIns.c:1506:17: runtime error: index 8 out of bounds for type 'uint8_t [8]'
    #0 0x55b85858136c in j__udyInsWalk third_party/judy/src/JudyCommon/JudyIns.c:1506:17
    #1 0x55b85857ce7f in j__udyInsWalk third_party/judy/src/JudyCommon/JudyIns.c:1650:23
    #2 0x55b85857ce7f in j__udyInsWalk third_party/judy/src/JudyCommon/JudyIns.c:1650:23
    #3 0x55b85857473d in Judy1Set third_party/judy/src/JudyCommon/JudyIns.c:1847:23
    #4 0x55b85812113f in bm::HandleMgr::get_handle(unsigned long*) third_party/p4lang_behavioral_model/include/bm/bm_sim/handle_mgr.h:186:5
    #5 0x55b8581140f4 in bm::ActionProfile::add_member(bm::ActionFn const*, bm::ActionData, unsigned int*) third_party/p4lang_behavioral_model/src/bm_sim/action_profile.cpp:215:21
    #6 0x55b85816a91f in bm::Context::mt_act_prof_add_member(std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, bm::ActionData, unsigned int*) third_party/p4lang_behavioral_model/src/bm_sim/context.cpp:157:20
    #7 0x55b857ea69d7 in bm::SwitchWContexts::mt_act_prof_add_member(unsigned int, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, bm::ActionData, unsigned int*) third_party/p4lang_behavioral_model/include/bm/bm_sim/switch.h:399:32
    #8 0x55b858036207 in _pi_act_prof_mbr_create third_party/p4lang_behavioral_model/PI/src/pi_act_prof_imp.cpp:112:38
    #9 0x55b858030ad8 in pi_act_prof_mbr_create third_party/p4lang_PI/src/pi_act_prof.c:32:10
    #10 0x55b8580295e8 in pi::ActProf::member_create(pi::ActionData const&, unsigned long*) third_party/p4lang_PI/frontends_extra/cpp/src/tables.cpp:771:10
    #11 0x55b857faad16 in pi::fe::proto::ActionProfAccessOneshot::group_create(p4::v1::ActionProfileActionSet const&, unsigned long*, pi::fe::proto::common::SessionTemp*) third_party/p4lang_PI/proto/frontend/src/action_prof_mgr.cpp:1133:27
    #12 0x55b857f878b5 in pi::fe::proto::DeviceMgrImp::construct_action_entry_oneshot(unsigned int, p4::v1::ActionProfileActionSet const&, pi::ActionEntry*, pi::fe::proto::common::SessionTemp*) third_party/p4lang_PI/proto/frontend/src/device_mgr.cpp:2584:5
    #13 0x55b857f847fc in pi::fe::proto::DeviceMgrImp::table_insert(p4::v1::TableEntry const&, pi::fe::proto::common::SessionTemp*) third_party/p4lang_PI/proto/frontend/src/device_mgr.cpp:2678:7
    #14 0x55b857f80811 in pi::fe::proto::DeviceMgrImp::table_write(p4::v1::Update_Type, p4::v1::TableEntry const&, pi::fe::proto::common::SessionTemp*) third_party/p4lang_PI/proto/frontend/src/device_mgr.cpp:691:16
    #15 0x55b857f57ebc in pi::fe::proto::DeviceMgrImp::write_(p4::v1::WriteRequest const&) third_party/p4lang_PI/proto/frontend/src/device_mgr.cpp:2112:20
    #16 0x55b857f5273b in pi::fe::proto::DeviceMgrImp::write(p4::v1::WriteRequest const&) third_party/p4lang_PI/proto/frontend/src/device_mgr.cpp:660:12
    #17 0x55b857f49dc5 in pi::fe::proto::DeviceMgr::write(p4::v1::WriteRequest const&) third_party/p4lang_PI/proto/frontend/src/device_mgr.cpp:3111:16
    #18 0x55b857f3736d in pi::server::(anonymous namespace)::P4RuntimeServiceImpl::Write(grpc::ServerContext*, p4::v1::WriteRequest const*, p4::v1::WriteResponse*) third_party/p4lang_PI/proto/server/pi_server.cpp:424:31
```

**Comment by smolkaj** [comment]
These are the offending lines for part of the stack trace:
```
// #4 bm::HandleMgr::get_handle(unsigned long*) third_party/p4lang_behavioral_model/include/bm/bm_sim/handle_mgr.h:186:5
J1S(Rc, handles, jhandle);  // Judy1Set() 

// #5 bm::ActionProfile::add_member(bm::ActionFn const*, bm::ActionData, unsigned int*) third_party/p4lang_behavioral_model/src/bm_sim/action_profile.cpp:215:21
if (mbr_handles.get_handle(&mbr_)) {

// #6 bm::Context::mt_act_prof_add_member(std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, bm::ActionData, unsigned int*) third_party/p4lang_behavioral_model/src/bm_sim/context.cpp:157:20
return act_prof->add_member(action, std::move(action_data), mbr);

// #7 bm::SwitchWContexts::mt_act_prof_add_member(unsigned int, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, bm::ActionData, unsigned int*) third_party/p4lang_behavioral_model/include/bm/bm_sim/switch.h:399:32
return contexts.at(cxt_id).mt_act_prof_add_member(
        act_prof_name, action_name, std::move(action_data), mbr);

// #8 _pi_act_prof_mbr_create third_party/p4lang_behavioral_model/PI/src/pi_act_prof_imp.cpp:112:38
  auto error_code = pibmv2::switch_->mt_act_prof_add_member(
      0, ap_name, a_name, std::move(adata), &h);
```

**Comment by antoninbas** [comment]
@smolkaj I think the Judy library may not be compiled correctly for your system. You may want to review your Bazel build file for libJudy. The library uses different pieces of code based on whether the system you will be running it on is 32-bit or 64-bit. This information needs to be provided at compile-time using the `JU_64BIT` macro.

A few years back Tom Everman had some choice words about libJudy: https://github.com/p4lang/PI/issues/46 :)

This is my Bazel build file for libJudy: https://github.com/p4lang/PI/blob/main/bazel/external/judy.BUILD. I am not claiming it doesn't suffer from the same issue though. But I believe they use it in the Stratum project and it's likely that's it been tested with more than 256 members (this seems to be roughly when things break and makes sense based on the info from ASAN). You may want to compare it to yours, notably with respect to how `JU_64BIT` is defined.

**Comment by smolkaj** [comment]
Thanks for the pointers. We're running on 64 bit. The JU_64BIT macro seems to be set correctly.

I tried applying the patch mentioned [here](https://sourceforge.net/p/judy/bugs/25/) and [here](https://sourceforge.net/p/judy/bugs/28/), and that makes the sanitizer happy, but BMv2 still fails with the same error.

**Comment by antoninbas** [comment]
I'm pretty convinced this is an issue with libJudy. It can come from the bazel build, the compiler or be a bug in the library itself. Unfortunately the library is very complex and it's clear that it's not actively maintained by the original author. I see that these folks are trying to maintain it: https://github.com/netdata/libjudy/commits/master. However it seems that the patch that they applied is one of the 2 you mention above. I am a bit puzzled by the fact that you get the same error after applying the patch with no more error from ASAN. Based on the stacktrace you shared the memory error in libJudy and the error you observe definitely seemed connected.

I think based on the consideration above, it would be a good idea to look into an alternative to libJudy. For the handle manager, `dynamic_bitset` from the boost library could be a good candidate. It will not perform as well as Judy (AFAIK) for large tables or action profiles, but should still perform ok. The worst case scenario is when the array is very sparse. If there are other libraries / data structures that someone would recommend, I'm open to other alternatives. It can be an alternative implementation at first, and the default one eventually.

Replacing libJudy in the LPM implementation may be a bit trickier but maybe it doesn't suffer from the same issue since it is a different array type... We should still strive to replace it though.

**Comment by smolkaj** [comment]
Just to be clear, I'm also onboard that Judy is the likely suspect here.
I'm wondering if there are any short-term workarounds I might try.

**Comment by antoninbas** [comment]
You can try out https://github.com/p4lang/behavioral-model/tree/support-non-judy-implementation-for-handle-mgr. It's a first stab at replacing the Judy-based implementation of the HandleMgr with something else (in this case boost/dynamic_bitset). You will need to provide the  `boost/dynamic_bitset.hpp` header in your bazel build.

**Comment by smolkaj** [comment]
Wow, amazing. This does the trick :) 
Thank you so much for taking the time to go through the lengthy debugging process with me, having the key idea to running the test with ASAN enabled, and then delivering a near-instant Judy replacement patch.

**Comment by smolkaj** [comment]
This issue can be closed as far as I am concerned, thanks again.

**Comment by antoninbas** [comment]
No problem. I'll keep working on the replacement when I have some more time. I am likely to drop the boost dependency altogether, so it should simplify the build for you. I'll close the issue when I have something merged into the main branch.

**Comment by antoninbas** [comment]
Closing this since libJudy is no longer used in bmv2 & PI
